File: client/lib/items/recipes.dart
/// Recipe table and the pure crafting rules — twin of `items/recipes.ts`. By
/// hand you can only make planks, sticks, a workbench and a wooden pickaxe; a
/// workbench within reach makes everything except another workbench.
library;

import 'package:block_survival/items/inventory.dart';

final class Recipe {
  const Recipe(this.id, this.output, this.inputs, {required this.bench});

  final String id;
  final ItemStack output;
  final List<ItemStack> inputs;
  final bool bench;
}

const List<Recipe> recipes = [
  Recipe('planks', ItemStack('planks', 4), [ItemStack('log', 1)], bench: false),
  Recipe('stick', ItemStack('stick', 4), [
    ItemStack('planks', 2),
  ], bench: false),
  Recipe('workbench', ItemStack('workbench', 1), [
    ItemStack('planks', 4),
  ], bench: false),
  Recipe('torch', ItemStack('torch', 4), [
    ItemStack('stick', 1),
    ItemStack('coal', 1),
  ], bench: true),
  Recipe('pickaxe_wood', ItemStack('pickaxe_wood', 1), [
    ItemStack('planks', 3),
    ItemStack('stick', 2),
  ], bench: false),
  Recipe('pickaxe_stone', ItemStack('pickaxe_stone', 1), [
    ItemStack('cobble', 3),
    ItemStack('stick', 2),
  ], bench: true),
  Recipe('pickaxe_copper', ItemStack('pickaxe_copper', 1), [
    ItemStack('copper', 3),
    ItemStack('stick', 2),
  ], bench: true),
  Recipe('pickaxe_iron', ItemStack('pickaxe_iron', 1), [
    ItemStack('iron', 3),
    ItemStack('stick', 2),
  ], bench: true),
  Recipe('pickaxe_gold', ItemStack('pickaxe_gold', 1), [
    ItemStack('gold', 3),
    ItemStack('stick', 2),
  ], bench: true),
  Recipe('pickaxe_diamond', ItemStack('pickaxe_diamond', 1), [
    ItemStack('diamond', 3),
    ItemStack('stick', 2),
  ], bench: true),
  Recipe('sword', ItemStack('sword', 1), [
    ItemStack('iron', 2),
    ItemStack('stick', 1),
  ], bench: true),
  Recipe('sword_diamond', ItemStack('sword_diamond', 1), [
    ItemStack('diamond', 2),
    ItemStack('stick', 1),
  ], bench: true),
  // issue #25: the lens that finds ore, and the emerald that widens its reach
  Recipe('prospector', ItemStack('prospector', 1), [
    ItemStack('lapis', 2),
    ItemStack('redstone', 2),
    ItemStack('iron', 1),
  ], bench: true),
  Recipe('prospector_far', ItemStack('prospector_far', 1), [
    ItemStack('prospector', 1),
    ItemStack('emerald', 2),
  ], bench: true),
  Recipe('glass', ItemStack('glass', 4), [
    ItemStack('sand', 4),
    ItemStack('coal', 1),
  ], bench: true),
  Recipe('reinforced_wall', ItemStack('reinforced_wall', 4), [
    ItemStack('cobble', 4),
    ItemStack('iron', 1),
  ], bench: true),
  Recipe('bed', ItemStack('bed', 1), [
    ItemStack('planks', 3),
    ItemStack('leaves', 3),
  ], bench: true),
  Recipe('rifle', ItemStack('rifle', 1), [
    ItemStack('iron', 8),
    ItemStack('planks', 2),
    ItemStack('coal', 2),
  ], bench: true),
  Recipe('ammo', ItemStack('ammo', 30), [
    ItemStack('iron', 1),
    ItemStack('coal', 1),
  ], bench: true),
  // redstone burns hotter than coal: the same iron goes twice as far
  Recipe('ammo_redstone', ItemStack('ammo', 60), [
    ItemStack('iron', 1),
    ItemStack('redstone', 1),
  ], bench: true),
];

enum CraftContext { hand, bench }

/// what the crafting panel lists: hand recipes anywhere, everything but the bench at a bench
List<Recipe> recipesFor(CraftContext context) => context == CraftContext.hand
    ? recipes.where((r) => !r.bench).toList()
    : recipes.where((r) => r.id != 'workbench').toList();

Recipe getRecipe(String id) => recipes.firstWhere(
  (r) => r.id == id,
  orElse: () => throw ArgumentError.value(id, 'id', 'unknown recipe'),
);

enum CraftStatus { ok, missing, needsBench }

CraftStatus craftStatus(
  Inventory inv,
  Recipe recipe, {
  required bool nearBench,
}) {
  if (recipe.bench && !nearBench) return CraftStatus.needsBench;
  return recipe.inputs.every((i) => inv.count(i.id) >= i.count)
      ? CraftStatus.ok
      : CraftStatus.missing;
}

/// Consume the inputs and add the output; the count that did not fit is
/// returned so the caller can drop it. Null when the recipe cannot be crafted.
int? craft(Inventory inv, Recipe recipe, {required bool nearBench}) {
  if (craftStatus(inv, recipe, nearBench: nearBench) != CraftStatus.ok) {
    return null;
  }
  for (final i in recipe.inputs) {
    inv.remove(i.id, i.count);
  }
  return inv.add(recipe.output.id, recipe.output.count);
}

File: node-server/inertia/items/recipes.ts
/**
 * Recipe table (GAME_PROMPT.md §5) and the pure crafting rules.
 * By hand (E, anywhere) you can only make planks, sticks, a workbench and a wooden
 * pickaxe; a Workbench within reach (F) makes everything except another workbench
 * (issue #10). `craftStatus` is the one rule the host enforces; `recipesFor` is what
 * the panel lists.
 */
import type { Inventory, ItemStack } from './inventory.ts'

export interface Recipe {
  id: string
  output: ItemStack
  inputs: readonly ItemStack[]
  bench: boolean
}

const r = (id: string, output: [string, number], inputs: [string, number][], bench: boolean): Recipe => ({
  id,
  output: { id: output[0], count: output[1] },
  inputs: inputs.map(([iid, count]) => ({ id: iid, count })),
  bench,
})

export const RECIPES: readonly Recipe[] = [
  r('planks', ['planks', 4], [['log', 1]], false),
  r('stick', ['stick', 4], [['planks', 2]], false),
  r('workbench', ['workbench', 1], [['planks', 4]], false),
  r('torch', ['torch', 4], [['stick', 1], ['coal', 1]], true),
  r('pickaxe_wood', ['pickaxe_wood', 1], [['planks', 3], ['stick', 2]], false),
  r('pickaxe_stone', ['pickaxe_stone', 1], [['cobble', 3], ['stick', 2]], true),
  r('pickaxe_copper', ['pickaxe_copper', 1], [['copper', 3], ['stick', 2]], true),
  r('pickaxe_iron', ['pickaxe_iron', 1], [['iron', 3], ['stick', 2]], true),
  r('pickaxe_gold', ['pickaxe_gold', 1], [['gold', 3], ['stick', 2]], true),
  r('pickaxe_diamond', ['pickaxe_diamond', 1], [['diamond', 3], ['stick', 2]], true),
  r('sword', ['sword', 1], [['iron', 2], ['stick', 1]], true),
  r('sword_diamond', ['sword_diamond', 1], [['diamond', 2], ['stick', 1]], true),
  // issue #25: the thing that finds ore. The first one is whittled out of wood, so
  // finding your first iron does not depend on already having mined it; lapis and
  // redstone, then emerald, widen its reach from there.
  r('prospector', ['prospector', 1], [['planks', 4], ['stick', 2]], true),
  r('prospector_tuned', ['prospector_tuned', 1], [['prospector', 1], ['lapis', 2], ['redstone', 2], ['iron', 1]], true),
  r('prospector_far', ['prospector_far', 1], [['prospector_tuned', 1], ['emerald', 2]], true),
  r('glass', ['glass', 4], [['sand', 4], ['coal', 1]], true),
  r('reinforced_wall', ['reinforced_wall', 4], [['cobble', 4], ['iron', 1]], true),
  r('bed', ['bed', 1], [['planks', 3], ['leaves', 3]], true),
  r('rifle', ['rifle', 1], [['iron', 8], ['planks', 2], ['coal', 2]], true),
  r('ammo', ['ammo', 30], [['iron', 1], ['coal', 1]], true),
  // redstone burns hotter than coal: the same iron goes twice as far
  r('ammo_redstone', ['ammo', 60], [['iron', 1], ['redstone', 1]], true),
]

export type CraftContext = 'hand' | 'bench'

/** what the crafting panel lists: hand recipes anywhere, everything but the bench at a bench */
export function recipesFor(context: CraftContext): readonly Recipe[] {
  return context === 'hand' ? RECIPES.filter(r => !r.bench) : RECIPES.filter(r => r.id !== 'workbench')
}

export const getRecipe = (id: string): Recipe => {
  const rec = RECIPES.find(x => x.id === id)
  if (!rec) throw new Error(`unknown recipe: ${id}`)
  return rec
}

export type CraftStatus = 'ok' | 'missing' | 'needsBench'

export function craftStatus(inv: Inventory, recipe: Recipe, nearBench: boolean): CraftStatus {
  if (recipe.bench && !nearBench) return 'needsBench'
  return recipe.inputs.every(i => inv.count(i.id) >= i.count) ? 'ok' : 'missing'
}

/**
 * Consume the inputs and add the output. Output that does not fit is returned so the
 * caller can drop it on the ground. Returns null if the recipe cannot be crafted.
 */
export function craft(inv: Inventory, recipe: Recipe, nearBench: boolean): { overflow: number } | null {
  if (craftStatus(inv, recipe, nearBench) !== 'ok') return null
  for (const i of recipe.inputs) inv.remove(i.id, i.count)
  const overflow = inv.add(recipe.output.id, recipe.output.count)
  return { overflow }
}
